### PR TITLE
A few changes to `dask.delayed`

### LINKED ISSUE #795: Inspect function annotations to enable unpacking of results with dask.imperative's do?
Unpacking is a really nice feature for writing readable Python code. Unfortunately, it can't work with dask imperative, because we don't have any way of determining the number of elements to iterate over.

But if we use function annotations for type hints, we could make something like the following work on Python 3:
```
def f(x: int) -> (int, int):
    return (x, x ** 2)

x, y = do(f)(123)
```

Currently this results in `TypeError: Value objects are not iterable`.

### PR DESCRIPTION
**Pull request by jcrist**
This makes a few separate changes to `dask.delayed`:

- Adds an optional `nout` keyword to `delayed` objects, allowing for
  splatting and iteration over results.  If not provided, iteration will
  still error. Fixes #795.

  ```python
  @delayed(pure=True, nout=2)
  def foo(a):
      return a, -a

  x, y = foo(1)
  ```

- Cleanup method calling behavior through use of small `DelayedAttr`
  object. Use `dask.utils.methodcaller` to do the actual method calls.
  This is cleaner than trying to inline the method getattrs later.

- Cleanup the test suite for delayed, removing a few duplicate tests.

- Remove long since deprecated `dask.imperative` and `dask.imperative.Value` namings.

### CODE REVIEW

File: dask/delayed.py [hunk 1]
```diff
@@ -292,24 +304,22 @@ class Delayed(base.Base):
 
     Equivalent to the output from a single key in a dask graph.
     """
-    __slots__ = ('_key', '_dasks')
+    __slots__ = ('_key', '_dasks', '_n')
```
Comment (mrocklin) [inline dask/delayed.py]: Perhaps a more informative name than `_n`?

File: dask/delayed.py [hunk 2]
```diff
@@ -367,52 +391,63 @@ def _get_binary_operator(cls, op, inv=False):
...
 
-        dasks = flat_unique(dasks)
-        dasks.append({name: task})
-        return Delayed(name, dasks)
+    @property
+    def _dasks(self):
+        return [{self._key: (getattr, self._obj._key, self._attr)}] + self._obj._dasks
+
+    def __call__(self, *args, **kwargs):
+        return call_function(methodcaller(self._attr), (self._obj,) + args, kwargs)
```
Comment (mrocklin) [inline dask/delayed.py]: Why the need for the extra class?

Comment (jcrist) [inline dask/delayed.py]: Makes the graphs resulting from method and attribute access cleaner. Also removes the need for the inlining optimization, as the graphs are created properly the first time. I think the class is simple enough to merit the addition.

Comment (mrocklin) [inline dask/delayed.py]: OK

**Comment by mrocklin** [comment]
+1

**Pull request merged**